File: Programmers/[17683] 방금그곡/Johoseong/17683.py
def solution(m, musicinfos):
    answer = "(None)"
    m = m.replace('C#', 'H').replace('D#', 'I').replace('F#', 'J').replace('G#', 'K').replace('A#', 'L') # 샾 붙은 애들 구별하려고 다른 알파벳으로 바꿈
    t = -1

    for i in musicinfos:
        s, e, title, melody = i.split(',')
        melody = melody.replace('C#', 'H').replace('D#', 'I').replace('F#', 'J').replace('G#', 'K').replace('A#', 'L')
        l = len(melody)
        tmp1, tmp2 = s.split(':'), e.split(':')

        time = (int(tmp2[0]) * 60 + int(tmp2[1])) - (int(tmp1[0]) * 60 + int(tmp1[1])) # 해당곡이 재생된 시간
        full_melody = (time // l) * melody + melody[:(time % l) + 1] # 재생된 시간만큼 멜로디 연결

        if (m in full_melody) and (t < time): # 재생된 멜로디 안에 m 포함됨 (더 오래 재생된게 answer)
            answer = title
            t = time

    return answer

File: Programmers/[17683] 방금그곡/KingDonggyu/17683.js
function replaceSharp(melody, music) {
  const targets = [
    [/C#/g, 'c'],
    [/D#/g, 'd'],
    [/F#/g, 'f'],
    [/G#/g, 'g'],
    [/A#/g, 'a'],
  ];

  targets.forEach((target) => {
    melody = melody.replace(target[0], target[1]);
    music = music.replace(target[0], target[1]);
  });

  return [melody, music];
}

function getPlayTime(start, end) {
  const minutes = [start, end].map((time) => {
    const splited = time.split(':');
    return +splited[0] * 60 + +splited[1];
  });

  return minutes[1] - minutes[0];
}

function checkSameMusic(melody, music, playTime) {
  if (playTime < music.length) {
    music = music.slice(0, playTime);
  } else {
    while (playTime > music.length) {
      music += music;
    }
  }

  return music.includes(melody);
}

function solution(m, musicinfos) {
  const answer = { title: null, playTime: 0 };

  musicinfos.forEach((musicinfo) => {
    const [start, end, title, music] = musicinfo.split(',');
    const replaced = replaceSharp(m, music);
    const playTime = getPlayTime(start, end);

    if (!checkSameMusic(...replaced, playTime)) {
      return;
    }

    if (answer.playTime < playTime) {
      answer.title = title;
      answer.playTime = playTime;
    }
  });

  if (!answer.title) {
    return '(None)';
  }

  return answer.title;
}
